comictalker/comictalker.py

```diff
@@ -0,0 +1,222 @@
+    ...
+
+
+def url_fetch_complete(image_url: str, thumb_url: str | None) -> None:
+    ...
+
+
+class ComicTalker:
+    alt_url_list_fetch_complete = list_fetch_complete
+    url_fetch_complete = url_fetch_complete
+
+    def __init__(self, source_name) -> None:
+        # ID of the source to use e.g. comicvine
+        self.source = source_name
+        # Retrieve the available sources modules
+        self.sources = self.get_talkers()
+        # Set the active talker
+        self.talker = self.get_active_talker()
+        self.static_options = self.get_static_options()
+
+    def get_active_talker(self):
+        # This should always work because it will have errored at get_talkers if there are none
+        if not self.sources[self.source] is None:
+            return self.sources[self.source]
+
+    @staticmethod
+    def get_talkers():
+        # Hardcode import for now. Placed here to prevent circular import
+        import comictalker.talkers.comicvine
+
+        return {"comicvine": comictalker.talkers.comicvine.ComicVineTalker()}
+
+    # For issueidentifier
+    def set_log_func(self, log_func: Callable[[str], None]) -> None:
+        self.talker.log_func = log_func
+
+    def get_static_options(self) -> SourceStaticOptions:
+        return self.talker.source_details.static_options
+
+    def check_api_key(self, key: str, url: str, source_id: str):
+        for source in self.sources.values():
+            if source.source_details.id == source_id:
+                return source.check_api_key(key, url)
+        # Return false as back up or error?
+        return False
+
+    # Master function to search for series/volumes
+    def search_for_series(
+        self,
+        series_name: str,
+        callback: Callable[[int, int], None] | None = None,
+        refresh_cache: bool = False,
+        literal: bool = False,
+    ) -> list[ComicVolume]:
+        try:
+            series_result = self.talker.search_for_series(series_name, callback, refresh_cache, literal)
+            return series_result
+        except NotImplementedError:
+            logger.warning(f"{self.talker.source_details.name} has not implemented: 'search_for_series'")
+            raise TalkerError(
+                self.talker.source_details.name,
+                4,
+                "The source has not implemented: 'search_for_series'",
+            )
+
+    # Master function to fetch series data (i.e for sources without issue level details)
+    def fetch_volume_data(self, series_id: int) -> GenericMetadata:
+        try:
+            series_result = self.talker.fetch_volume_data(series_id)
+            return series_result
+        except NotImplementedError:
+            logger.warning(f"{self.talker.source_details.name} has not implemented: 'fetch_volume_data'")
+            raise TalkerError(
+                self.talker.source_details.name,
+                4,
+                "The source has not implemented: 'fetch_volume_data'",
+            )
```

Comment:
Can we combine this with `fetch_issue_data`? Is there any reason why we can't make a single function `fetch_comic_data` and if there is no issue_number set only get the "volume" data

Reply:
That does bring up the question of fetch_issue_data_by_issue_id. That could be added something like:
```
def fetch_comic_data(self, series_id: int, issue_number: str = "", issue_id: int = 0) -> GenericMetadata:
if issue_id:
  ...
elif self.source_details.static_options.has_issues and issue_number:
  ...
else:
  # Only retrieve the volume data
```

Reply:
Yeah that's possible but looking at where it would be used, it adds more complexity than usefulness in my current opinion.
If you look at where `fetch_issue_data` and `fetch_volume_data` were used it was, does it support issues? Yes: Return the volume and issue data. No: Return just the volume data. If you look a little closer even with your merge the difference between `fetch_volume_data` and `fetch_partial_volume_data` is a list of fields to get from comicvine.

`fetch_issue_data_by_issue_id` on the other hand is only used in one place. The function signature from `fetch_issue_data` `fetch_comic_data` doesn't even change and with setting the `issue_number` parameter to a default value even `fetch_volume_data` doesn't really change either. The change from `fetch_issue_data_by_issue_id` to `fetch_comic_data` is a murkier change because it's not clear what to do with the parameter.

My opinion on this may change by the time this merges.

Reply:
@mizaki my opinion changed 😂.
These should be merged, this cuts down on having a separate function to implement that is incredibly similar.
You can still keep the function internally but making someone implement two functions (that are almost the same) is worse than having three parameters that can be cleared up by a decent docstring.